### ISSUE TITLE
Support Emacs 27.1

### ISSUE THREAD

**Issue by asr** [agda/agda#4610]
Emacs 27.0.91 (the second pretest for what will be the 27.1) was [announced](https://lists.gnu.org/archive/html/emacs-devel/2020-04/msg01123.html). I'm creating this issue for collecting/reporting the required changes/issues.

Initially, I got the following errors:

```
$ agda-mode compile
In toplevel form:
.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/annotation.el:6:1:Error: Package cl is deprecated

In toplevel form:
.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/eri.el:7:1:Error: Package cl is deprecated

In toplevel form:
.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda-input.el:22:1:Error: Package cl is deprecated

In toplevel form:
.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda2-highlight.el:11:1:Error: Package cl is deprecated

In toplevel form:
.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda2-mode.el:17:1:Error: Package cl is deprecated
Unable to compile the following Emacs Lisp files:
  /home/asr/.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/annotation.el
  /home/asr/.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/eri.el
  /home/asr/.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda-input.el
  /home/asr/.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda2-highlight.el
  /home/asr/.cabal/share/x86_64-linux-ghc-8.10.1/Agda-2.6.2/emacs-mode/agda2-mode.el
```

I know Agda developers using different versions of GHC, which is good. I would appreciate if you tell us which Emacs version are you using.

**Comment by jespercockx** [comment]
I'm using GNU Emacs 25.2.2

**Comment by gallais** [comment]
I have GNU Emacs 24.5.1 installed.

**Comment by ajrouvoet** [comment]
GNU Emacs 26.3

4 developers, 4 versions of Emacs

**Comment by fredrikNordvallForsberg** [comment]
(another GNU Emacs 25.2.2 here)

**Comment by guillaumebrunerie** [comment]
I’m using GNU Emacs 25.2.2 as well (but I’m on Ubuntu 18.04 LTS which will presumably get upgraded to the new 20.04 LTS with a brand new Emacs somewhat soon).

**Comment by nad** [comment]
GNU Emacs 25.2.2.

**Comment by andreasabel** [comment]
Emacs 26.1 (Mac).

**Comment by L-TChen** [comment]
Another GNU Emacs 26.3 (Mac) installed via homebrew.

**Comment by mouse07410** [comment]
GNU Emacs 27.0.91 (Mac) installed via Macports. Same error as shown above.

**Comment by dalvescb** [comment]
GNU Emacs 27.0.91 Arch Linux installed via AUR emacs27-git. Is there a workaround for this or just need to downgrade emacs??

**Comment by asr** [comment]
As workaround I guess you can remove the line https://github.com/agda/agda/blob/58710f312ea66c1da0ee9f8a1fe44be9d7883d2e/src/agda-mode/Main.hs#L234

**Comment by ice1000** [comment]
I'm using a dev version of Emacs: GNU Emacs 28.0.50 under WSL.

Same error.

**Comment by xekoukou** [comment]
GNU Emacs 26.1 (Debian)

**Comment by tonyfloatersu** [comment]
GNU Emacs 27.1 Arch Linux here, same problem : /

**Comment by asr** [comment]
I removed my assignment to this issue.

**Comment by L-TChen** [comment]
The package `cl` is deprecated in favour of `cl-lib` in Emacs 27.

**Comment by tonyfloatersu** [comment]
Thanks! I think its resolved for me.

**Comment by L-TChen** [comment]
@tonyfloatersu Would you like to contribute your solution? I am working on it, but my Emacs Lisp literacy is rather poor.

**Comment by nad** [comment]
Can the issues be fixed by replacing `(require 'cl)` by `(require 'cl-lib)` (for Emacs 24.3 and later)?

**Comment by L-TChen** [comment]
Not really. You will get the following error message if you use `cl-lib` without modification:
```
In toplevel form:
.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/annotation.el:287:22:Error: reference to free variable ‘end’

In toplevel form:
.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/eri.el:207:1:Error: the function ‘position-if’ is not known to be defined.

In toplevel form:
.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/agda-input.el:55:30:Error: ‘(f1 f)’ is a malformed function

In toplevel form:
.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/agda2-mode.el:24:1:Error: Symbol’s function definition is void: lexical-let
Unable to compile the following Emacs Lisp files:
  /Users/liang-tingchen/Projects/agda/.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/annotation.el
  /Users/liang-tingchen/Projects/agda/.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/eri.el
  /Users/liang-tingchen/Projects/agda/.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/agda-input.el
  /Users/liang-tingchen/Projects/agda/.stack-work/install/x86_64-osx/a76264dcdd21b0ea9a86b8ad9918f3ad3061e6339e36501982f294e7346f29cb/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/emacs-mode/agda2-mode.el
```

**Comment by L-TChen** [comment]
Alright, I think I know what to do. I will try to get this done later today.

**Comment by tonyfloatersu** [comment]
Yeah I am also working on it for now. Sorry about the wrong message, I forgot to check `agda-mode compile`. My apologies.
As @nad and @L-TChen said, replacing `(require 'cl)` with `(require 'cl-lib)` in `emacs-mode` directory doesn't help much.

Just did a little bit of investigation: 
- At least the error in `eri.el` can be solved by replacing `position-if` with `cl-position-if`. It seems they are the same by construction in `cl-lib`.
- Complaint in `agda-input.el` (also brought to `agda2-mode.el`) is caused by `lexical-let` in `cl`.
  (`cl-lib` doesn't have it.)
  Maybe we want to try `lexical-binding: t` to be a fix? Here's a [link](https://www.reddit.com/r/emacs/comments/5j9qgz/lexicalbinding_t_vs_lexicallet/) might be helpful (?)

p.s. I am really a noob in Elisp, plz correct me if I say anything doesn't make sense.

**Comment by wenkokke** [comment]
@L-TChen I opened a pull-request which replaces cl with cl-lib, and moves all calls to cl symbols to their cl-lib equivalents: <https://github.com/agda/agda/pull/4894>. 

However, I might've moved symbols that don't use any Common Lisp features: cl overrides a lot of macros, such as defun and defmacro, and I didn't check to see if the cl features were truly needed.

**Comment by L-TChen** [comment]
@wenkokke I have opened a PR #4886 for this. (This issue was referenced but I did not leave a comment explicitly here. Sorry for the duplicate work.)

**Comment by wenkokke** [comment]
@L-TChen No worries! I quickly patched up the emacs-mode to get it to work on my machine, and figured I might as well submit a PR.

**Comment by andreasabel** [comment]
Fixed by @L-TChen in PR #4886 commit aa5e3a127bf17a8c80d947f3c286758a36dadc36.
